internal/lifecycle/readme.go

```diff
@@ -65,19 +118,24 @@ func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
 		return nil, errNoREADMECodeBlocksFound
 	}
 
-	lifecycle, err := codeBlocksTolifecycle(codeBlocks, serviceName, gcrURL)
-	if err != nil {
-		return lifecycle, fmt.Errorf("[lifecycle.parseREADME] transforming code blocks in %s to executable commands: %w", filename, err)
+	var l Lifecycle
+	for _, b := range codeBlocks {
+		c, err := b.toCommands(serviceName, gcrURL)
+		if err != nil {
+			return l, fmt.Errorf("[lifecycle.parseREADME] transforming code blocks in %s to executable commands: %w", filename, err)
+		}
+
+		l = append(l, c...)
```

Comment:
Looking at this logic, I think that properly speaking you should redefine `NewLifecycle` to take a slice of codeBlock as a means of initializing, so you always have the same entry point. It breaks the encapsulation somewhat to have a NewLifecycle function but then have other ways of constructing a Lifecycle as well. The surface would be something like:

```
// Surfaces:
type LifecycleOption interface {
  apply(l *Lifecycle) // this method applies the option to the lifecycle.
}

type withCodeBlocks struct {
  cb []CodeBlock
}

func (w withCodeBlocks) apply(l *Lifecycle) {
  // this does the logic that you have in L122-129. 
}

func WithCodeBlocks(blocks []CodeBlock) LifecycleOption {
  // returns a withCodeBlocks struct using the given blocks.
}

// additional functions to produce appropriate lifecycle options as needed.

func NewLifecycle(opts ...LifecycleOption)  // This could have additional params such as service name if needed.

// To use:
l := NewLifecycle(FromCodeBlocks(blocks))
```

I don't think you have to fix this in order to merge this PR, but this is the kind of pattern that I would expect to see in order to polish this up.

Reply:
Discussed this offline with Saketram-- I totally got the logic backwards here (since actually NewLifecycle is what calls the readme parsing logic), sorry about that!

I think it would be preferable for NewLifecycle to have options specifying what kind of lifecycle is desired rather than using defaults based on the environment or what files are present, but I'm not sure if that's fully compatible with project requirements.